alf/environments/alf_gym3_wrapper.py

```diff
@@ -0,0 +1,415 @@
+        """Construct an adapted instance for the input Gym3 environment
+
+        Args:
+
+            gym3_env: the input environment which should be an instance of a
+                class that derives from gym3.Env
+            image_channel_first: when set to True, the image-based (of 3
+                channels) observation will be permuted so that the channel
+                dimension comes first.
+            ignored_info_keys: a list of keys in the env info that should not be
+                included in the env info of the TimeStep. This is useful when
+                some huge but not useful information are stored in the env info
+                of the underlying Gym3 environment, and ignoring them is crucial
+                to achieve better performance.
+            support_force_reset: Gym3 environments do not support force reset in
+                general. However, some of the environments such as procgen
+                allows sending action -1 to reset the environments. Set this to
+                True to enable such behavior.
+            render_activator: when set to None, it indicates that this
+                environment does not support rendering. Otherwise it will be a
+                function that re-creates a Gym3 environment with render enabled.
+                See render() for details.
+            frame_extractor: when set to None, it indicates that this
+                environment does not support recording. Otherwise it will be a
+                function that extracts the rendered frame for recording from the
+                environment.
+
+        """
+        assert isinstance(gym3_env, gym3.Env), \
+            f'AlfGym3Wrapper: {type(gym3_env)} is not dervied from gym3.Env'
+        super().__init__()
+
+        # The underlying Gym3 environment
+        self._gym3_env = gym3_env
+
+        self._support_force_reset = support_force_reset
+
+        # +--------------------------+
+        # | Render/Recording Related |
+        # +--------------------------+
+
+        # When initially constructed, render is not enabled until the first call
+        # to render() is invoked. Use self._render_enabled to make sure render
+        # is not enabled for more than once.
+        self._render_enabled = False
+        self._render_activator = render_activator
+        self._frame_extrator = frame_extractor
+        # Create metadata with 'render.modes' so that it is compatible with
+        # VideoRecorder.
+        self.metadata = {'render.modes': []}
+        if self._render_activator is not None and self._frame_extrator is not None:
+            self.metadata['render.modes'].append('rgb_array')
+
+        # +--------------------------+
+        # | Cache the Tensor Specs   |
+        # +--------------------------+
+
+        # NOTE(breakds): when needed, expose this and allow an user to set it.
+        self._discount = torch.full((self.batch_size, ), 1.0)
```

Comment:
Will procgen games have normal episode ends like success or failure or game-over? If so, then in those cases, self._discount should be 0.

Reply:
Yep that is pretty much what `_done` is doing. Thanks a lot for pointing this out - I originally thought `step type` will have full information of this, but that is not used in computing advantages/accumulated returns.

Reply:
Renamed back to `_prev_first`.